### ISSUE TITLE
Add remove_cdrom to vmware-iso

### ISSUE THREAD

**Issue by hc-github-team-packer** [hashicorp/packer-plugin-vmware#33]
_This issue was originally opened by @jeremymcgee73 in https://github.com/hashicorp/packer/issues/10704 and has been migrated to this repository. The original issue description is below._

---

#### Community Note

Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request.
Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request.
If you are interested in working on this issue or have submitted a pull request, please leave a comment.

#### Description

I would like to be able to use remove_cdrom with the vmware-iso provider. I am adding a disk using cd_files with vmware-iso, then I am use vmware-vmx to import the vmx, do some more config and convert to ova. However, the conversion fails because the temporary ISO no longer exists. I would like to cleanup the CD drive in the vmware-iso run. I may be able to fix this with ovftool_options, but have not gotten that to work yet.

#### Potential configuration

```
```

**Comment by azr** [comment]
I think this should have been fixed by #30 & #31, so, resolving; please open a new issue if you still have this ! 🙂

**Comment by puetzk** [comment]
#31 restores the cleanup that removes the temporary ISO and turns it into a CDROM device with no media present. This should fix the OVA export error, but it would still have a CDROM device present (packer doesn't really keep track of whether it added the device, or just temporarily switched it to have a .iso mounted)

So it depends on whether @jeremymcgee73 specifically wanted the cdrom device removed, or just wanted the .ova export to not be broken by dangling temp files.

**Comment by azr** [comment]
Good point, reopening in case.

**Comment by puetzk** [comment]
If we really wanted to make this cleanup more aggressive, it would probably be simplest to make `temporaryDevices` a dictionary (rather than a list), which saves the original values (or nil for absence?), so StepCleanVMX could just put the vmx fill back however it originally was (delete the key if we added it, revert its value if it was present but changed value).

**Comment by arizvisa** [comment]
Yeah, Packer used to set the cdrom device to non-present or something similar after the build. Now, if you do an install from an iso the generated .vmx will reference said iso. This means tools like `ovftool` believe that the iso (that was only used to do the install) is related to the template when really it was only needed to perform the install and isn't needed anymore.

**Comment by arizvisa** [comment]
Okay, was doing a bisect but it turned out to be the latest commit related to thse things. It's commit 048e31aeb6d810f2683dd6a17ab867c3508a8982 which happened in August. Inside iso/step_create_vmx.go, it used to add the cdrom device to the list of `temporaryDevices` since the cdrom device was being used for the install.

```
	/// Now that we figured out the CDROM device to add, store it
	/// to the list of temporary build devices in our statebag
	tmpBuildDevices := state.Get("temporaryDevices").([]string)
	tmpCdromDevice := fmt.Sprintf("%s0:%s", templateData.CDROMType, templateData.CDROMType_PrimarySecondary)
	tmpBuildDevices = append(tmpBuildDevices, tmpCdromDevice)
	state.Put("temporaryDevices", tmpBuildDevices)
```

The author noted the following with regards to removing the installation cdrom drive from the list of temporary devices. 
```
    * Track  cdrom in temporaryDevices for step_clean_vmx

    There was a previous attempt to do this in step_configure_vmx,
    but that only covers vmware-iso (missing potential cases when using
    cd_files/cd_content in vmware-vmx) and it had the wrong device name
    (would mark <type>0:<primary> as temporary when the device name used
    for cd_path is actually <type>1:<primary>
```

**Comment by puetzk** [comment]
Ah, you're right. The `if` you added doesn't really make sense, but the unconditional addition in StepCreateISO that seemed to me an erroneous attempt to remove the temporary for `cd_files` 
https://github.com/hashicorp/packer-plugin-vmware/blob/fea55ff74505c7e9b0224554e1a87e5fc6ffe3a8/builder/vmware/iso/step_create_vmx.go#L173-L179

was actually doing something else entirely - it was balancing out a device which was not "added" through `vmx_data[cdromPrefix+...]`, but rather written into the original template:
https://github.com/hashicorp/packer-plugin-vmware/blob/274f4049159f7ada33bd0742c72c0ef27fee59d7/builder/vmware/iso/step_create_vmx.go#L485-L487

But that "original template" was itself created out of thin air by packer, so those are indeed meant to be as temporary as things added later via StepConfigureVMX modifying the file. And that's why it was "{CDROMType}0:{PrimarySecondary}" instead of "{CDROMType}1:{PrimarySecondary}".

Mea culpa. So yes, these [lines I mistakenly removed](https://github.com/arizvisa/hashicorp-packer-plugin-vmware/commit/048e31aeb6d810f2683dd6a17ab867c3508a8982#diff-69b312abca15e1a6e28abae7463a3d2ce360e110d54a0b2cf9d37bf3429f7e3aL173-L179) should indeed just be put back (as they were, not wrapped in any if based on CDFiles or CDContent like your 40d123cb05de0edca88d613fcf3acebea08a81af). They weren't the "wrong device name", they just never had anything to do with the temporary cdrom in StepConfigureVMX. And it would be legit to have both `iso_target_path` and `cd_files`, and you'd have two temporary CDROM devices, and cleanup should remove the `cdrom-image` from both.

**Comment by arizvisa** [comment]
Ah. Okay. Yeah, it was prolly easier to understand for me because I was the original author. Heh. ;)

I have no idea how the cd_contents and related actually work, tho. I was thinking that cd_files would be a way to detect if a user is using a cd device for something else and that they'd want it to persist through the cleaning step. Nonetheless, if you patch it w/ a single commit, could you note that commit in this thread?

I'm not sure if that was what the OP actually desired..but the symptoms are very similar and if this feature is \_not\_ a result of that issue, then I imagine that there'll be feature requests to allow users to remove or disable the cdrom drives through some other means as a result of the mentioned change.

**Comment by puetzk** [comment]
> I have no idea how the cd_contents and related actually work, tho

packer calls mkisofs, and makes a temporary .iso. Usually this is to stick in [kickstart](https://access.redhat.com/documentation/en-us/red_hat_enterprise_linux/7/html/installation_guide/sect-kickstart-howto)/[preseed](https://wiki.debian.org/DebianInstaller/Preseed)/[Unattend](https://docs.microsoft.com/en-us/windows-hardware/manufacture/desktop/update-windows-settings-and-scripts-create-your-own-answer-file-sxs) kinds of files, or maybe something like [NoCloud](https://cloudinit.readthedocs.io/en/latest/topics/datasources/nocloud.html) metadata if you're starting from a .ova or .vmx that already has cloud-init in it. Packer's going to delete that .iso at the end of the run, so those are also ephemeral and should be unmounted from the .vmx. If you wanted to create a CDROM device that would last, I guess you could do that with vmx_data, but packer doesn't have a specific feature for it.

I thought this code was trying to clean up from the temporarily created .iso and just got the device name ide0:0 vs ide1:0 wrong somehow. But just actually potentially have two IDE cdrom devices (one form vmware-iso's `iso_target_path` managing name ide0:0 and one from cd_{files,content} managing name ide1:0 (or other names if you didn't pick IDE or made them slave devicss). You could have one, both, or neither. And that's fine - they're using different decide IDs, and having two CDROMs may be a little odd but it's allowed (somewhere I have a even have old physical PC tower like that, since it had a CD burner and later got a DVD reader added).

**Comment by arizvisa** [comment]
ah. thx for the clarification.

**Comment by tenthirtyam** [comment]
Resolved in https://github.com/hashicorp/packer-plugin-vmware/pull/42.

cc @nywilken 


**Ryan Johnson**
Staff II Solutions Architect | VMware, Inc.

**Comment by nywilken** [comment]
Closing as this issue has been resolved by release [1.0.5](https://github.com/hashicorp/packer-plugin-vmware/releases/tag/v1.0.5)
